maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,204 @@
+        self._event_buffer.insert_event(vm_required_evt)
+
+    def get_metrics(self) -> dict:
+        """Get current enviornment metrics information.
+
+        Args:
+            dict: A dict that contains the environment metrics.
+
+        Returns:
+            dict: Metrics information.
+        """
+
+        energy_consumption = self._energy_consumption
+        success_requirements = self._success_requirements
+        total_latency = self._total_latency
+        return DocableDict(
+            metrics_desc,
+            energy_consumption=energy_consumption,
+            success_requirements=success_requirements,
+            total_latency=total_latency
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(Events.REQUIREMENTS, self._on_vm_required)
+        self._event_buffer.register_event_handler(Events.FINISHED, self._on_vm_finished)
+
+        # Decision event
+        self._event_buffer.register_event_handler(MaroEvents.TAKE_ACTION, self._on_action_received)
+
+    def _update_vm_util(self, tick: int):
+        """Update all live VMs memory utilization."""
+        for _, vm in self.vm.items():
+            vm.util_mem = vm.show_util_series()[tick]
+
+    def _update_pm_util(self, tick: int):
+        """Update memory utilization occupied by total VMs on each PM."""
+        for pm in self.machines:
+            cur_total_mem: int = 0
+            for vm_id in pm.show_vm_set():
+                vm = self.vm[vm_id]
+                vm.util_mem = vm.show_util_series()[tick]
+                cur_total_mem += vm.util_mem
+            pm.util_mem = cur_total_mem
+            pm.update_util_series(cur_total_mem)
+
+    def _on_vm_required(self, evt: CascadeEvent):
+        """Callback when there is a VM requirement generated."""
+        # Load VM data
+        payload: Payload = evt.payload
+        vm_req: VirtualMachine = payload.vm_info
+        buffer_time: int = payload.buffer_time
+
+        if any([(pm.cap_mem - pm.util_mem) >= vm_req.req_mem for pm in self.machines]):
+            # Generate pending decision
+            pending_decision_evt = self._event_buffer.gen_decision_event(evt.tick, payload=payload)
+            evt.add_immediate_event(pending_decision_evt)
+        else:
+            # Postpone to next tick
+            if buffer_time > 0:
+                postpone_payload = deepcopy(payload)
+                postpone_payload.buffer_time -= 1
+                self._total_latency.resource_buffer_time += 1
+                postpone_evt = self._event_buffer.gen_cascade_event(evt.tick + 1, payload=postpone_payload)
+                self._event_buffer.insert_event(postpone_evt)
+            else:
+                # Fail
+                pass
+
+    def _on_vm_finished(self, evt: AtomEvent):
+        """Callback when there is a VM in the end cycle."""
+        # Remove dead VM
+        vm_id: int = evt.payload
```

Comment:
better to define a payload class for this event. it will be more unified in the code and more scalable if we need to add more features.